core/shared/src/main/scala/zio/ZEnvironment.scala

```diff
@@ -214,8 +214,11 @@ final class ZEnvironment[+R] private (
   def unionAll[R1](that: ZEnvironment[R1]): ZEnvironment[R with R1] =
     if (self.relaxedEquals(that)) that.asInstanceOf[ZEnvironment[R with R1]]
     else {
-      val newMap = that.map.iterator.foldLeft(self.map) { case (map, (k, v)) =>
-        map.updated(k, v)
+      var newMap = self.map
+      val it     = that.map.iterator
+      while (it.hasNext) {
+        val (k, v) = it.next()
+        newMap = newMap.updated(k, v)
```

Comment:
Why can't we use `addAll`?

Reply:
The map that we use here is a custom internal one that does not extend Scala's `immutable.Map` so `addAll` doesn't exist. Although I think we could add that method since I think we could definitely optimize it for batch additions.

I'll do that in a separate PR though as I think it's a bit outside the scope (pun intended) of this PR